clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java

```diff
@@ -456,6 +457,36 @@ protected Map<Node, FetchSessionHandler.FetchRequestData> prepareFetchRequests()
         return fetchable.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().build()));
     }
 
+    private Set<Integer> nodesWithBufferedPartitions(Set<TopicPartition> buffered, long currentTimeMs) {
+        Set<Integer> nodesWithBufferedPartitions = new HashSet<>();
+
+        for (TopicPartition partition : buffered) {
+            if (!subscriptions.isAssigned(partition)) {
+                // It's possible that a partition with buffered data from a previous request is now no longer
+                // assigned to the consumer, in which case just skip this partition.
+                continue;
+            }
+
+            SubscriptionState.FetchPosition position = subscriptions.position(partition);
```

Comment:
In nodesWithBufferedPartitions, we skip if position is null or if the leader is empty. 

Previously, we would throw illegal state exception (L409) or log and request a metadata update for buffered partitions.

Is there a reason for this divergence of this, or should we merge the two logic together?

Reply:
The thinking was that in that case, potentially sending a fetch request to a node with buffered data was preferable to throwing an exception. But that may not be the right call. I'll look at refactoring the logic.

Reply:
I see we have duplicate code to find the node for buffered & unbuffered. can we fix that?

Reply:
Code has been refactored. I went through three or four different ways of refactoring, most of which ended up changing more stuff in the _existing_ fetch code than I liked. The refactoring I did end up with is very minimal. PTAL.